File: native/physics2d/framework/component/WheelJoint2DComponent.cpp
#include "WheelJoint2DComponent.h"
#include "../PhysicsSystem.h"
#include "../../../engine/core/core.h"

using namespace physics2d;

void WheelJoint2DComponent::onSwitchConnectNode()
{
    tempDef = onCreateJointDef();
    setBody(this->node, connectNode, *tempDef);
    joint = PhysicsSystem::getInstance()->getWorld()->getPyxhWorld()->createWheelJoint(*((phxy::SqWheelJointDef *)tempDef));
    delete tempDef;
    tempDef = nullptr;
}

phxy::SqBaseJointDef *WheelJoint2DComponent::onCreateJointDef()
{
    if (!tempDef)
        tempDef = new phxy::SqWheelJointDef();
    return tempDef;
}

void WheelJoint2DComponent::setEnableSpring(float enable)
{
    if (joint)
    {

        dynamic_cast<phxy::SqWheelJoint *>(joint)->EnableSpring(enable);
    }
    else
    {
        tempDef = onCreateJointDef();
        ((phxy::SqWheelJointDef *)tempDef)->enableSpring = enable;
    }
}

void WheelJoint2DComponent::setHertz(float v)
{
    if (joint)
    {
        dynamic_cast<phxy::SqWheelJoint *>(joint)->SetSpringHertz(v);
    }
    else
    {
        tempDef = onCreateJointDef();
        ((phxy::SqWheelJointDef *)tempDef)->hertz = v;
    }
}

void WheelJoint2DComponent::setDampingRatio(float v)
{
    if (joint)
    {
        dynamic_cast<phxy::SqWheelJoint *>(joint)->SetSpringDampingRatio(v);
    }
    else
    {
        tempDef = onCreateJointDef();
        ((phxy::SqWheelJointDef *)tempDef)->dampingRatio = v;
    }
}

void WheelJoint2DComponent::setEnableLimit(bool enable)
{
    if (joint)
    {
        dynamic_cast<phxy::SqWheelJoint *>(joint)->EnableLimit(enable);
    }
    else
    {
        tempDef = onCreateJointDef();
        ((phxy::SqWheelJointDef *)tempDef)->enableLimit = enable;
    }
}

void WheelJoint2DComponent::setLowerTranslation(float t)
{
    if (joint)
    {
        phxy::SqWheelJoint *wjoint = dynamic_cast<phxy::SqWheelJoint *>(joint);
        wjoint->SetLimits(t, wjoint->GetUpperLimit());
    }
    else
    {
        tempDef = onCreateJointDef();
        ((phxy::SqWheelJointDef *)tempDef)->lowerTranslation = t;
    }
}

void WheelJoint2DComponent::setUpperTranslation(float t)
{
    if (joint)
    {
        phxy::SqWheelJoint *wjoint = dynamic_cast<phxy::SqWheelJoint *>(joint);
        wjoint->SetLimits(wjoint->GetLowerLimit(), t);
    }
    else
    {
        tempDef = onCreateJointDef();
        ((phxy::SqWheelJointDef *)tempDef)->upperTranslation = t;
    }
}

void WheelJoint2DComponent::setEnableMotor(bool enable)
{
    if (joint)
    {
        dynamic_cast<phxy::SqWheelJoint *>(joint)->EnableMotor(enable);
    }
    else
    {
        tempDef = onCreateJointDef();
        ((phxy::SqWheelJointDef *)tempDef)->enableMotor = enable;
    }
}

void WheelJoint2DComponent::setMaxMotorTorque(float v)
{
    if (joint)
    {
        dynamic_cast<phxy::SqWheelJoint *>(joint)->SetMaxMotorTorque(v);
    }
    else
    {
        tempDef = onCreateJointDef();
        ((phxy::SqWheelJointDef *)tempDef)->maxMotorTorque = v;
    }
}

void WheelJoint2DComponent::setMotorSpeed(float v)
{
    v = Math::angleToRadian(v);
    if (joint)
    {
        dynamic_cast<phxy::SqWheelJoint *>(joint)->SetMotorSpeed(v);
    }
    else
    {
        tempDef = onCreateJointDef();
        ((phxy::SqWheelJointDef *)tempDef)->motorSpeed = v;
    }
}

File: sq/pyhsics-2d/framework/component/WheelJoint2DComponent.ts
import { serializable, SQBoolean, sqclass, type, SQFloat } from "../../../core/index";
import { Joint2DComponent } from "./Joint2DComponent";
import { ComponentType } from "../../../native_binding/index";


@sqclass("sq.WheelJoint2DComponent")
export default class WheelJoint2DComponent extends Joint2DComponent {
    private _enableSpring?: boolean;
    private _hertz?: number;
    private _dampingRatio?: number;
    private _enableLimit?: boolean;
    private _lowerTranslation?: number;
    private _upperTranslation?: number;
    private _enableMotor?: boolean;
    private _maxMotorTorque?: number;
    private _motorSpeed?: number;

    constructor() {
        super();
        this.nativeType = ComponentType.WheelJoint2DComponent;
    }


    @type(SQBoolean)
    @serializable
    set enableSpring(enable: boolean) {
        if (this._enableSpring !== enable) {
            this._enableSpring = enable;
            this.nativeBeginOp(51);
            this.nativeWriteOpArg("i8", enable ? 1 : 0);
            this.nativeEndOp();
        }
    }
    get enableSpring() {
        return this._enableSpring;
    }

    @type(SQFloat)
    @serializable
    set hertz(v: number) {
        if (this._hertz !== v) {
            this._hertz = v;
            this.nativeBeginOp(52);
            this.nativeWriteOpArg("f32", v);
            this.nativeEndOp();
        }
    }
    get hertz() {
        return this._hertz;
    }

    @type(SQFloat)
    @serializable
    set dampingRatio(v: number) {
        if (this._dampingRatio !== v) {
            this._dampingRatio = v;
            this.nativeBeginOp(53);
            this.nativeWriteOpArg("f32", v);
            this.nativeEndOp();
        }
    }

    get dampingRatio() {
        return this._dampingRatio;
    }

    @type(SQBoolean)
    @serializable
    set enableLimit(enable: boolean) {
        if (this._enableLimit !== enable) {
            this._enableLimit = enable;
            this.nativeBeginOp(54);
            this.nativeWriteOpArg("i8", enable ? 1 : 0);
            this.nativeEndOp();
        }
    }

    get enableLimit() {
        return this._enableLimit;
    }

    @type(SQFloat)
    @serializable
    set lowerTranslation(angle: number) {
        if (this._lowerTranslation !== angle) {
            this._lowerTranslation = angle;
            this.nativeBeginOp(55);
            this.nativeWriteOpArg("f32", angle);
            this.nativeEndOp();
        }
    }
    get lowerTranslation() {
        return this._lowerTranslation;
    }

    @type(SQFloat)
    @serializable
    set upperTranslation(angle: number) {
        if (this._upperTranslation !== angle) {
            this._upperTranslation = angle;
            this.nativeBeginOp(56);
            this.nativeWriteOpArg("f32", angle);
            this.nativeEndOp();
        }
    }

    get upperTranslation() {
        return this._upperTranslation;
    }

    @type(SQBoolean)
    @serializable
    set enableMotor(enable: boolean) {
        if (this._enableMotor !== enable) {
            this._enableMotor = enable;
            this.nativeBeginOp(57);
            this.nativeWriteOpArg("i8", enable ? 1 : 0);
            this.nativeEndOp();
        }
    }
    get enableMoto() {
        return this._enableMotor;
    }

    @type(SQFloat)
    @serializable
    set maxMotorTorque(v: number) {
        if (this._maxMotorTorque !== v) {
            this._maxMotorTorque = v;
            this.nativeBeginOp(58);
            this.nativeWriteOpArg("f32", v);
            this.nativeEndOp();
        }
    }

    get maxMotorTorque() {
        return this._maxMotorTorque;
    }

    @type(SQFloat)
    @serializable
    set motorSpeed(v: number) {
        if (this._motorSpeed !== v) {
            this._motorSpeed = v;
            this.nativeBeginOp(59);
            this.nativeWriteOpArg("f32", v);
            this.nativeEndOp();
        }
    }
    get motorSpeed() {
        return this._motorSpeed;
    }
}